File: py/brasil_sdk.py
# Brasil SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import BrasilUtility
from core.spec import BrasilSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import BrasilBaseFeature
from features import _make_feature


class BrasilSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = BrasilUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return BrasilUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = BrasilSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def bank(self):
        """Idiomatic facade: client.bank.list() / client.bank.load({"id": ...})."""
        from entity.bank_entity import BankEntity
        cached = getattr(self, "_bank", None)
        if cached is None:
            cached = BankEntity(self, None)
            self._bank = cached
        return cached

    def Bank(self, data=None):
        # Deprecated: use client.bank instead.
        from entity.bank_entity import BankEntity
        return BankEntity(self, data)


    @property
    def cep(self):
        """Idiomatic facade: client.cep.list() / client.cep.load({"id": ...})."""
        from entity.cep_entity import CepEntity
        cached = getattr(self, "_cep", None)
        if cached is None:
            cached = CepEntity(self, None)
            self._cep = cached
        return cached

    def Cep(self, data=None):
        # Deprecated: use client.cep instead.
        from entity.cep_entity import CepEntity
        return CepEntity(self, data)


    @property
    def cnpj(self):
        """Idiomatic facade: client.cnpj.list() / client.cnpj.load({"id": ...})."""
        from entity.cnpj_entity import CnpjEntity
        cached = getattr(self, "_cnpj", None)
        if cached is None:
            cached = CnpjEntity(self, None)
            self._cnpj = cached
        return cached

    def Cnpj(self, data=None):
        # Deprecated: use client.cnpj instead.
        from entity.cnpj_entity import CnpjEntity
        return CnpjEntity(self, data)


    @property
    def ddd(self):
        """Idiomatic facade: client.ddd.list() / client.ddd.load({"id": ...})."""
        from entity.ddd_entity import DddEntity
        cached = getattr(self, "_ddd", None)
        if cached is None:
            cached = DddEntity(self, None)
            self._ddd = cached
        return cached

    def Ddd(self, data=None):
        # Deprecated: use client.ddd instead.
        from entity.ddd_entity import DddEntity
        return DddEntity(self, data)


    @property
    def feriado(self):
        """Idiomatic facade: client.feriado.list() / client.feriado.load({"id": ...})."""
        from entity.feriado_entity import FeriadoEntity
        cached = getattr(self, "_feriado", None)
        if cached is None:
            cached = FeriadoEntity(self, None)
            self._feriado = cached
        return cached

    def Feriado(self, data=None):
        # Deprecated: use client.feriado instead.
        from entity.feriado_entity import FeriadoEntity
        return FeriadoEntity(self, data)


    @property
    def fipe_marca(self):
        """Idiomatic facade: client.fipe_marca.list() / client.fipe_marca.load({"id": ...})."""
        from entity.fipe_marca_entity import FipeMarcaEntity
        cached = getattr(self, "_fipe_marca", None)
        if cached is None:
            cached = FipeMarcaEntity(self, None)
            self._fipe_marca = cached
        return cached

    def FipeMarca(self, data=None):
        # Deprecated: use client.fipe_marca instead.
        from entity.fipe_marca_entity import FipeMarcaEntity
        return FipeMarcaEntity(self, data)


    @property
    def fipe_preco(self):
        """Idiomatic facade: client.fipe_preco.list() / client.fipe_preco.load({"id": ...})."""
        from entity.fipe_preco_entity import FipePrecoEntity
        cached = getattr(self, "_fipe_preco", None)
        if cached is None:
            cached = FipePrecoEntity(self, None)
            self._fipe_preco = cached
        return cached

    def FipePreco(self, data=None):
        # Deprecated: use client.fipe_preco instead.
        from entity.fipe_preco_entity import FipePrecoEntity
        return FipePrecoEntity(self, data)


    @property
    def municipio(self):
        """Idiomatic facade: client.municipio.list() / client.municipio.load({"id": ...})."""
        from entity.municipio_entity import MunicipioEntity
        cached = getattr(self, "_municipio", None)
        if cached is None:
            cached = MunicipioEntity(self, None)
            self._municipio = cached
        return cached

    def Municipio(self, data=None):
        # Deprecated: use client.municipio instead.
        from entity.municipio_entity import MunicipioEntity
        return MunicipioEntity(self, data)


    @property
    def ufn(self):
        """Idiomatic facade: client.ufn.list() / client.ufn.load({"id": ...})."""
        from entity.ufn_entity import UfnEntity
        cached = getattr(self, "_ufn", None)
        if cached is None:
            cached = UfnEntity(self, None)
            self._ufn = cached
        return cached

    def Ufn(self, data=None):
        # Deprecated: use client.ufn instead.
        from entity.ufn_entity import UfnEntity
        return UfnEntity(self, data)


    @property
    def ufn2(self):
        """Idiomatic facade: client.ufn2.list() / client.ufn2.load({"id": ...})."""
        from entity.ufn2_entity import Ufn2Entity
        cached = getattr(self, "_ufn2", None)
        if cached is None:
            cached = Ufn2Entity(self, None)
            self._ufn2 = cached
        return cached

    def Ufn2(self, data=None):
        # Deprecated: use client.ufn2 instead.
        from entity.ufn2_entity import Ufn2Entity
        return Ufn2Entity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: rb/Brasil_sdk.rb
# Brasil SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Brasil_types'


class BrasilSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = BrasilUtility.new
    @_utility = utility

    config = BrasilConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = BrasilHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = BrasilHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, BrasilFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    BrasilUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = BrasilHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = BrasilHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = BrasilHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = BrasilSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue BrasilError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = BrasilHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = BrasilHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.bank.list / client.bank.load({ "id" => ... })
  def bank
    require_relative 'entity/bank_entity'
    @bank ||= BankEntity.new(self, nil)
  end

  # Deprecated: use client.bank instead.
  def Bank(data = nil)
    require_relative 'entity/bank_entity'
    BankEntity.new(self, data)
  end


  # Idiomatic facade: client.cep.list / client.cep.load({ "id" => ... })
  def cep
    require_relative 'entity/cep_entity'
    @cep ||= CepEntity.new(self, nil)
  end

  # Deprecated: use client.cep instead.
  def Cep(data = nil)
    require_relative 'entity/cep_entity'
    CepEntity.new(self, data)
  end


  # Idiomatic facade: client.cnpj.list / client.cnpj.load({ "id" => ... })
  def cnpj
    require_relative 'entity/cnpj_entity'
    @cnpj ||= CnpjEntity.new(self, nil)
  end

  # Deprecated: use client.cnpj instead.
  def Cnpj(data = nil)
    require_relative 'entity/cnpj_entity'
    CnpjEntity.new(self, data)
  end


  # Idiomatic facade: client.ddd.list / client.ddd.load({ "id" => ... })
  def ddd
    require_relative 'entity/ddd_entity'
    @ddd ||= DddEntity.new(self, nil)
  end

  # Deprecated: use client.ddd instead.
  def Ddd(data = nil)
    require_relative 'entity/ddd_entity'
    DddEntity.new(self, data)
  end


  # Idiomatic facade: client.feriado.list / client.feriado.load({ "id" => ... })
  def feriado
    require_relative 'entity/feriado_entity'
    @feriado ||= FeriadoEntity.new(self, nil)
  end

  # Deprecated: use client.feriado instead.
  def Feriado(data = nil)
    require_relative 'entity/feriado_entity'
    FeriadoEntity.new(self, data)
  end


  # Idiomatic facade: client.fipe_marca.list / client.fipe_marca.load({ "id" => ... })
  def fipe_marca
    require_relative 'entity/fipe_marca_entity'
    @fipe_marca ||= FipeMarcaEntity.new(self, nil)
  end

  # Deprecated: use client.fipe_marca instead.
  def FipeMarca(data = nil)
    require_relative 'entity/fipe_marca_entity'
    FipeMarcaEntity.new(self, data)
  end


  # Idiomatic facade: client.fipe_preco.list / client.fipe_preco.load({ "id" => ... })
  def fipe_preco
    require_relative 'entity/fipe_preco_entity'
    @fipe_preco ||= FipePrecoEntity.new(self, nil)
  end

  # Deprecated: use client.fipe_preco instead.
  def FipePreco(data = nil)
    require_relative 'entity/fipe_preco_entity'
    FipePrecoEntity.new(self, data)
  end


  # Idiomatic facade: client.municipio.list / client.municipio.load({ "id" => ... })
  def municipio
    require_relative 'entity/municipio_entity'
    @municipio ||= MunicipioEntity.new(self, nil)
  end

  # Deprecated: use client.municipio instead.
  def Municipio(data = nil)
    require_relative 'entity/municipio_entity'
    MunicipioEntity.new(self, data)
  end


  # Idiomatic facade: client.ufn.list / client.ufn.load({ "id" => ... })
  def ufn
    require_relative 'entity/ufn_entity'
    @ufn ||= UfnEntity.new(self, nil)
  end

  # Deprecated: use client.ufn instead.
  def Ufn(data = nil)
    require_relative 'entity/ufn_entity'
    UfnEntity.new(self, data)
  end


  # Idiomatic facade: client.ufn2.list / client.ufn2.load({ "id" => ... })
  def ufn2
    require_relative 'entity/ufn2_entity'
    @ufn2 ||= Ufn2Entity.new(self, nil)
  end

  # Deprecated: use client.ufn2 instead.
  def Ufn2(data = nil)
    require_relative 'entity/ufn2_entity'
    Ufn2Entity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = BrasilSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
